### ISSUE TITLE
1.67 regression with `……::{opaque#0}<'_> does not live long enough` error

### ISSUE THREAD

**Issue by steffahn** [rust-lang/rust#107516]
Discussed in https://users.rust-lang.org/t/rust-1-67-lifetime-error-compiles-with-1-66-what-changed/88483 and described on https://github.com/zotero/citeproc-rs/issues/154. I can reproduce a regression in 1.67.0 using the current latest commit “[`2ab195a1e6f84f0ff284813ece61dc62096abbfe`](https://github.com/zotero/citeproc-rs/commit/2ab195a1e6f84f0ff284813ece61dc62096abbfe)” in the `zotero/citeproc-rs` repo.

This issue was first incorrectly thought of as a sort of “duplicate” of #107426, since that issue originally claimed a 1.67 regression, too, but actually regressed in 1.66, and on a deliberate soundness fix (#95474).

However, bisection of the above-mentioned commit on `citeproc-rs` points to a different PR, that is
* #103491

Glancing at it, that PR does not look like it’s *supposed* to break anything. @rustbot label regression-from-stable-to-stable, T-compiler, E-needs-mcve

Compilation error
```
error: `QuoteSplitter<'_, impl Iterator<Item = (usize, char)> + '_>::events::{opaque#0}<'_>` does not live long enough
   --> crates/io/src/output/parse_quotes.rs:419:47
    |
419 |                 | MicroNode::Formatted(..) => EachSplitter::Index(Some(ix)),
    |                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `QuoteSplitter<'_, impl Iterator<Item = (usize, char)> + '_>::events::{opaque#0}<'_>` does not live long enough
   --> crates/io/src/output/parse_quotes.rs:430:21
    |
430 | /                     EachSplitter::Splitter {
431 | |                         index: ix,
432 | |                         splitter,
433 | |                         seen_any: Some(false),
434 | |                     }
    | |_____________________^

error: could not compile `citeproc-io` due to 2 previous errors
```

bisection
````
searched toolchains a28f3c88e50a77bc2a91889241248c4543854e61 through b7bc90fea3b441234a84b49fdafeb75815eebbab


********************************************************************************
Regression in 7fe6f36224e92db6fbde952e0b7e50863161f6ee
********************************************************************************

==================================================================================
= Please file this regression report on the rust-lang/rust GitHub repository     =
=        New issue: https://github.com/rust-lang/rust/issues/new                 =
=     Known issues: https://github.com/rust-lang/rust/issues                     =
= Copy and paste the text below into the issue report thread.  Thanks!           =
==================================================================================

searched nightlies: from nightly-2022-09-10 to nightly-2023-01-20
regressed nightly: nightly-2022-11-22
searched commit range: https://github.com/rust-lang/rust/compare/a28f3c88e50a77bc2a91889241248c4543854e61...b7bc90fea3b441234a84b49fdafeb75815eebbab
regressed commit: https://github.com/rust-lang/rust/commit/7fe6f36224e92db6fbde952e0b7e50863161f6ee

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.5</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --prompt --access github --start 2022-09-10 --end 2023-01-20 --regress error 
```
````

**Comment by steffahn** [comment]
@rustbot label A-impl-trait

Ping @cjgillot because of the regressing PR.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23107516.201.2E67.20regression.20with.20.60.E2.80.A6.E2.80.A6.3A.3A.7Bopaque.230.7D.3C'_.3E.60.20does.20not.20.E2.80.A6)).

@rustbot label -I-prioritize +P-critical

**Comment by pnkfelix** [comment]
(I'm planning to do a MCVE hunt for this after I take care of a beta- and stable-backport that is slightly higher priority.)

**Comment by steffahn** [comment]
Perhaps this is obvious, but to avoid any wasted time, I’ll rather write the remark than not: While reducing/minimizing – given the experience described in https://github.com/rust-lang/rust/issues/107426#issuecomment-1410729872 – make sure to keep an eye on that the regression point doesn’t shift :cat:

**Comment by compiler-errors** [comment]
Ah, guess I should note that @cjgillot has not been tagged here. I may work on MCVE here, but they know that PR best.

**Comment by compiler-errors** [comment]
Partially minimized:

```rust
#[derive(Debug)]
struct QuoteMatcher<'a>(&'a Vec<String>);

impl<'a> QuoteMatcher<'a> {
    fn intermediates(&'a self) -> impl Iterator<Item = Intermediate> + 'a {
        self.0.iter().enumerate().flat_map(move |(ix, string)| {
            let prev = self.0.get(ix.wrapping_sub(1)).and_then(|n| leaning_text(n, true));
            let next = self.0.get(ix + 1).and_then(|n| leaning_text(n, false));
            let splitter = new_quote_splitter(&string, prev, next).events();
            EachSplitter::Splitter { index: ix, splitter, seen_any: Some(false) }
        })
    }
}

impl<'a, I: Iterator<Item = (usize, char)>> QuoteSplitter<'a, I> {
    fn events(self) -> impl Iterator<Item = Event<'a>> {
        self.flat_map(|x: Thingo| x).filter(|ev| match ev {
            _ => true,
        })
    }
}

#[derive(Debug)]
enum EachSplitter<'a, I: Iterator<Item = Event<'a>> + 'a> {
    Splitter { splitter: I, seen_any: Option<bool>, index: usize },
}

impl<'a, I: Iterator<Item = Event<'a>>> Iterator for EachSplitter<'a, I> {
    type Item = Intermediate;
    fn next(&mut self) -> Option<Self::Item> {
        todo!()
    }
}

struct Event<'a>(&'a str);

struct Intermediate;

fn leaning_text(node: &str, rightmost: bool) -> Option<&str> {
    todo!()
}

#[derive(Debug)]
struct QuoteSplitter<'a, I: Iterator<Item = (usize, char)> + 'a> {
    string: &'a str,
    previous_text_node: Option<&'a str>,
    subsequent_text_node: Option<&'a str>,
    text_start: usize,
    possibles: I,
    emitted_last: bool,
}

fn new_quote_splitter<'a>(
    string: &'a str,
    prev: Option<&'a str>,
    next: Option<&'a str>,
) -> QuoteSplitter<'a, impl Iterator<Item = (usize, char)> + 'a> {
    QuoteSplitter {
        string,
        previous_text_node: prev,
        subsequent_text_node: next,
        text_start: 0,
        possibles: string.char_indices().filter(|_| true),
        emitted_last: false,
    }
}

impl<'a, I: Iterator<Item = (usize, char)>> Iterator for QuoteSplitter<'a, I> {
    type Item = Thingo<'a>;
    fn next(&mut self) -> Option<Self::Item> {
        todo!()
    }
}

fn main() {}

struct Thingo<'a>(&'a ());

impl<'a> Iterator for Thingo<'a> {
    type Item = Event<'a>;
    fn next(&mut self) -> Option<Self::Item> {
        todo!()
    }
}
```

**Comment by wesleywiser** [comment]
Would someone from @rust-lang/types be able to investigate this regression a bit? Thanks! 🙂

**Comment by aliemjay** [comment]
Minimized further:
```rust
fn iter<'a: 'a>() -> impl Iterator<Item = &'static str> {
    [].into_iter()
}

struct Bivar<'a, I: Iterator<Item = &'a str> + 'a>(I);

fn test() {
    || {
        Bivar(iter())
    };
}
```

**Comment by pnkfelix** [comment]
Hmm. I get errors from @aliemjay 's example even on old compilers, of the form:

```
  = help: the trait `Sized` is not implemented for `str`
  = note: required because of the requirements on the impl of `Iterator` for `std::slice::Iter<'_, str>`
```

The easiest way to sidestep this is to revise the supposed MCVE with one that uses a sized type, like `()` or `u32`, instead of `str`.

**Comment by jackh726** [comment]
@pnkfelix I was also having trouble, but all it takes is edition 2021

**Comment by pnkfelix** [comment]
Oh yeah, this is the Array::into_iter thing I bet, okay.

**Comment by aliemjay** [comment]
Fixed by #108691 (hopefully, because there are many ways for minimization to get wrong). It turned out fixing #108592 first was not necessary.

@rustbit claim
